test/observer/app.dart

```diff
@@ -0,0 +1,30 @@
+import 'package:flutter/services.dart';
+
+import '../test.dart';
+
+/// An observer for the application state. 
+class AppObserver {
+  
+  /// Whether a close request was recorded.
+  bool _haveCloseRequest = false;
+
+  /// Create a new application state observer, which automatically
+  /// unregisters any previously created observer.
+  AppObserver(WidgetTester tester) {
+    tester.binding.defaultBinaryMessenger.setMockMethodCallHandler(SystemChannels.platform, (call) {
+      if (call.method == 'SystemNavigator.pop') {
+        _haveCloseRequest = true;
+        return null;
+      }
+      return null;  // Ignore unimplemented method calls
+    });
+  }
+
+  /// Execute the provided [callable] and return whether
+  /// an application close request was made by it. 
+  Future<bool> haveCloseRequest(Future<void> Function() callable) async {
+    _haveCloseRequest = false;
+    await callable();
```

Comment:
This does not need to have a callable, this class can just expose the number of times SystemNavigator.pop was called, and then we can check that number in tests that need this

Reply:
The idea was to ensure that the request happened in the given callable and not before. I thought it would look better than having

```dart
appObserver.clearCloseRequests();
// Do something
expect(appObserver.hadCloseRequest(), true);
```

What do you think?

Reply:
Same reasoning for `expectShowsToast` and `expectShowsNoToast`.

Reply:
I can see the idea behind this, but I have never seen tests like this. It's more intuitive that we write

```dart
expect(systemChannelsObserver.popCount, 0);
callSomethingThatTriggersPop();
expect(systemChannelsObserver.popCount, 1);
```

Reply:
I opted for an approach where the getter clears the property on access. This is produces the cleanest test code, and should be robust enough against reordering of steps in the test.